File: pack/cpp/bitwidth.cpp
#include <stdint.h>


inline uint32_t bits(const uint32_t v) {
    return v == 0 ? 0 : 32 - __builtin_clz(v);
}

uint32_t maxbits(const uint32_t *in, const uint32_t size) asm("maxbits");

uint32_t maxbits(const uint32_t *in, const uint32_t size) {
    uint32_t acc = 0;
    for (int i = 0; i < size; i++) {
        acc |= in[i];
    }
    return bits(acc);
}

File: pack/bitwidth.go
package pack

import "math/bits"

func maxBits(in []uint32) uint32 {
	acc := uint32(0)
	for _, v := range in {
		acc |= v
	}
	return uint32(bits.Len64(uint64(acc)))
}

func maxBits2(in []uint32) uint32 {
	min := in[0]
	max := in[0]
	for _, v := range in {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
	}
	return uint32(bits.Len64(uint64(max - min)))
}
